### PR TITLE
Add --allowlist-file option

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: src/ir/item.rs [hunk 1]
```diff
@@ -530,6 +530,11 @@ impl Item {
         &mut self.kind
     }
 
+    /// Where in the source is this item located?
+    pub fn location(&self) -> &Option<clang::SourceLocation> {
```
Comment (emilio) [inline src/ir/item.rs:534]: Let's return an `Option<&clang::SourceLocation>` and use `self.location.as_ref()` instead, as that's a bit more idiomatic.

File: src/ir/context.rs [hunk 2]
```diff
@@ -2318,6 +2319,21 @@ If you encounter an error missing from this list, please file an issue or a PR!"
                         return true;
                     }
 
+                    // Items with a source location in an explicitly allowlisted file
+                    // are always included.
+                    if let Some(location) = &item.location() {
```
Comment (emilio) [inline src/ir/context.rs:2324]: That way you can remove this ampersand.

File: src/ir/context.rs [hunk 3]
```diff
@@ -2318,6 +2319,21 @@ If you encounter an error missing from this list, please file an issue or a PR!"
...
+                        let (file, _, _, _) = location.location();
+                        if let Some(filename) = file.name() {
```
Comment (emilio) [inline src/ir/context.rs:2326]: This does a fair amount of ffi calls and allocations, so let's check that `allowlisted_files` has something before bothering getting the filename.

**Review by emilio — CHANGES_REQUESTED** [review]
Thanks! Looks good with the nits addressed.

**Comment by sfackler** [comment]
This branch is working great for openssl-sys :) https://github.com/sfackler/rust-openssl/pull/1573

File: src/ir/item.rs (hunk 1 above)
Comment (daviddrysdale) [reply src/ir/item.rs:534]: Done.

File: src/ir/context.rs (hunk 2 above)
Comment (daviddrysdale) [reply src/ir/context.rs:2324]: Done.

File: src/ir/context.rs (hunk 3 above)
Comment (daviddrysdale) [reply src/ir/context.rs:2326]: Done. (There's a check for any kind of allowlist further up that I thought would be sufficient.)

**Comment by sfackler** [comment]
Is this ready to land?

**Comment by daviddrysdale** [comment]
> Is this ready to land?

Yes, as far as I know &ndash; @emilio, anything else needed/wanted?

**Comment by ianks** [comment]
We are trying to use this feature in https://github.com/ianks/rb-sys for official Ruby support for Rust. Would love to see this get merged!

**Comment by sfackler** [comment]
Any chance you could cut a release with this? Thanks!

**Pull request merged**